src/platform/plugins/shared/controls/public/controls/data_controls/types.ts

```diff
@@ -53,6 +53,14 @@ export interface DataControlFactory<
   CustomOptionsComponent?: React.FC<CustomOptionsComponentProps<State>>;
 }
 
+export interface DataOrESQLControlFactory<
```

Comment:
This is reveling a code smell with `DataControlFactory`. It should allow composition instead of inheriting `ControlFactory`. We should cleanup this typing but that can wait for another PR.

In the short term, lets remove  `DataOrESQLControlFactory` and just change `DataOrESQLControlFactory` state to extend `DefaultControlState` instead of `DefaultDataControlState`

```
export interface DataControlFactory<
  State extends DefaultControlState = DefaultControlState,
  Api extends DataControlApi = DataControlApi
> extends ControlFactory<State, Api> {
  isFieldCompatible: (field: DataViewField) => boolean;
  CustomOptionsComponent?: React.FC<CustomOptionsComponentProps<State>>;
}
```